File: Assets/src/generator/TileSection.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;
using UnityEngine;

[Serializable]
public class TileSection {
  public Type[,] Types { get; }
  public readonly string source;
  public TileSection(Type[,] types, string source) {
    Types = types;
    this.source = source;
  }

  public int width => Types.GetLength(0);
  public int height => Types.GetLength(1);
  public Vector2Int Size => new Vector2Int(width, height);

    public TileSection Rot90() {
    // rotate 90 degrees clockwise
    var newTypes = new Type[height, width];
    for (int x = 0; x < width; x++) {
      for (int y = 0; y < height; y++) {
        newTypes[y, width - 1 - x] = Types[x, y];
      }
    }
    return new TileSection(newTypes, source);
  }

  public void Blit(Floor floor, Vector2Int topLeft) {
    var bottomLeft = topLeft - new Vector2Int(0, height - 1);
    for (int x = 0; x < width; x++) {
      for (int y = 0; y < height; y++) {
        var pos = bottomLeft + new Vector2Int(x, y);
        if (!floor.InBounds(pos)) {
          continue;
        }

        var tileType = Types[x, y];
        var constructor = tileType.GetConstructor(new Type[] { typeof(Vector2Int) });
        var tile = (Tile)constructor.Invoke(new object[] { pos });
        floor.Put(tile);
      }
    }
  }

  public static TileSection FromString(string source) {
    /* s will look like: @"
xx_
xxx
 x
"
    */
    source = source.Trim();
    if (source.Length == 0) {
      throw new Exception("Empty string!");
    }
    var lines = source.Split(new string[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);

    // Text's coordinate system is +y down. Our game's coordinate system is +y up.
    // To account for this we flip the order of lines here.
    lines = lines.Reverse().ToArray();

    var width = lines.Select(line => line.Length).Max();
    var height = lines.Length;
    Type[,] Types = new Type[width, height];
    for(int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        var line = lines[y];
        var curChar = line[x];
        Type tileType = null;
        switch (curChar) {
          case 'x':
            tileType = typeof(Wall);
            break;
          case 'c':
            tileType = typeof(Chasm);
            break;
          case '_':
          default:
            tileType = typeof(Ground);
            break;
        }
        Types[x, y] = tileType;
      }
    }
    return new TileSection(Types, source);
  }

  public static TileSection[] FromMultiString(string multiSections) {
    var sections = multiSections.Split(new string[] { Environment.NewLine + Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
    return sections.Select(section => FromString(section.Trim())).ToArray();
  }

  internal static TileSection[] WithRotations(TileSection[] tileSections) {
    var rot90 = tileSections.Select(tileSection => tileSection.Rot90());
    var rot180 = rot90.Select(tileSection => tileSection.Rot90());
    var rot270 = rot180.Select(tileSection => tileSection.Rot90());
    return tileSections.Concat(rot90).Concat(rot180).Concat(rot270).ToArray();
  }
}

/**
  public static TileSection[] StandardSectionsSmall = FromMultiString(@"
xx_
xxx
_xx

xx
_x

_x_
xxx
_x_

_x_
xxx
__x

__x_
xxxx

xxxx
_x__

xxx
xx_

xx_
xxx

x__
xxx
x__
");

  public static TileSection[] StandardSections = FromMultiString(@"
xxx_
xxxx
xxx_
_xx_

_xxx_
xxxx_
_xxxx
__xx_

xxxx
xx_x
xxxx
_xxx

__xx
_xxx
xxxx
xxx_
xx__

_x_x_
xxxx_
_xxxx
__xxx

xx_xx
xxxx_
_xxxx
xx_xx

xx__
xxx_
xxxx
_xx_

====

xxx_
__xx
x_x_
xxxx

====

xxxx
x_xx
x__x
xxxx

====

xxxx_
x___x
x_x_x
xxxxx

====

xxx___
xxx___
___xxx
___xxx
___xx_

====

__xx__
_xxxx_
xxx_xx
xx__x_
_xxxx_

");
*/

File: web/src/generator/TileSection.ts
import { Vector2Int } from '../core/Vector2Int';
import type { Floor } from '../model/Floor';
import { Ground, Wall, Chasm } from '../model/Tile';

type TileFactory = (pos: Vector2Int) => any;

/**
 * Tile pattern template for stamping onto floors.
 * Port of C# TileSection.cs.
 */
export class TileSection {
  /** 2D array [x][y] of tile factory functions */
  readonly types: (TileFactory | null)[][];
  readonly source: string;

  constructor(types: (TileFactory | null)[][], source: string) {
    this.types = types;
    this.source = source;
  }

  get width(): number { return this.types.length; }
  get height(): number { return this.types[0]?.length ?? 0; }

  /** Rotate 90° clockwise */
  rot90(): TileSection {
    const w = this.width;
    const h = this.height;
    const newTypes: (TileFactory | null)[][] = [];
    for (let y = 0; y < h; y++) {
      newTypes[y] = [];
      for (let x = 0; x < w; x++) {
        newTypes[y][w - 1 - x] = this.types[x][y];
      }
    }
    return new TileSection(newTypes, this.source);
  }

  /** Stamp this pattern onto the floor, topLeft is the top-left corner in game coords */
  blit(floor: Floor, topLeft: Vector2Int): void {
    const bottomLeft = Vector2Int.sub(topLeft, new Vector2Int(0, this.height - 1));
    for (let x = 0; x < this.width; x++) {
      for (let y = 0; y < this.height; y++) {
        const pos = Vector2Int.add(bottomLeft, new Vector2Int(x, y));
        if (!floor.inBounds(pos)) continue;
        const factory = this.types[x][y];
        if (factory) {
          floor.put(factory(pos));
        }
      }
    }
  }

  /** Parse ASCII art into a TileSection. x=Wall, c=Chasm, _/default=Ground */
  static fromString(source: string): TileSection {
    const trimmed = source.trim();
    if (trimmed.length === 0) throw new Error('Empty TileSection string!');

    // Text is +y down; game is +y up — reverse line order
    const lines = trimmed.split('\n').map(l => l.trimEnd()).reverse();
    const height = lines.length;
    const width = Math.max(...lines.map(l => l.length));

    const types: (TileFactory | null)[][] = [];
    for (let x = 0; x < width; x++) {
      types[x] = [];
      for (let y = 0; y < height; y++) {
        const ch = x < lines[y].length ? lines[y][x] : '_';
        switch (ch) {
          case 'x': types[x][y] = (pos) => new Wall(pos); break;
          case 'c': types[x][y] = (pos) => new Chasm(pos); break;
          default: types[x][y] = (pos) => new Ground(pos); break;
        }
      }
    }
    return new TileSection(types, source);
  }

  /** Parse multiple sections separated by blank lines */
  static fromMultiString(source: string): TileSection[] {
    return source.split(/\n\s*\n/)
      .map(s => s.trim())
      .filter(s => s.length > 0)
      .map(s => TileSection.fromString(s));
  }

  /** Generate all 4 rotations for each section */
  static withRotations(sections: TileSection[]): TileSection[] {
    const result: TileSection[] = [];
    for (const s of sections) {
      const r90 = s.rot90();
      const r180 = r90.rot90();
      const r270 = r180.rot90();
      result.push(s, r90, r180, r270);
    }
    return result;
  }
}
